src/NHibernate/Linq/LinqExtensionMethods.cs

```diff
@@ -77,51 +81,81 @@ public static IQueryable<T> Timeout<T>(this IQueryable<T> query, int timeout)
+		/// </summary>
+		/// <param name="source">An <see cref="T:System.Linq.IQueryable`1" /> to convert to a future query.</param>
+		/// <typeparam name="TSource">The type of the elements of <paramref name="source" />.</typeparam>
+		/// <returns>A <see cref="IFutureValue{T}"/>.</returns>
+		/// <exception cref="T:System.ArgumentNullException"><paramref name="source" /> is <see langword="null"/>.</exception>
+		/// <exception cref="T:System.NotSupportedException"><paramref name="source" /> <see cref="IQueryable.Provider"/> is not a <see cref="INhQueryProvider"/>.</exception>
+		public static IFutureValue<TSource> ToFutureValue<TSource>(this IQueryable<TSource> source)
 		{
-			var nhQueryable = query as QueryableBase<T>;
-			if (nhQueryable == null)
-				throw new NotSupportedException("Query needs to be of type QueryableBase<T>");
+			if (source == null)
+			{
+				throw new ArgumentNullException(nameof(source));
+			}
+			if (!(source.Provider is INhQueryProvider provider))
+			{
+				throw new NotSupportedException($"Source {nameof(source.Provider)} must be a {nameof(INhQueryProvider)}");
+			}
+			var future = provider.ExecuteFuture<TSource>(source.Expression);
+			return new FutureValue<TSource>(() => future);
+		}
 
-			var provider = (INhQueryProvider) nhQueryable.Provider;
-			var future = provider.ExecuteFuture(nhQueryable.Expression);
-			if (future is IEnumerable<T>)
+		/// <summary>
+		/// Wraps the query in a deferred <see cref="IFutureValue{T}"/> which will trigger a batch of all pending future queries
+		/// when its <see cref="IFutureValue{T}.Value"/> is read.
+		/// </summary>
+		/// <param name="source">An <see cref="T:System.Linq.IQueryable`1" /> to convert to a future query.</param>
+		/// <param name="selector">An aggregation function to apply to <paramref name="source"/>.</param>
+		/// <typeparam name="TSource">The type of the elements of <paramref name="source" />.</typeparam>
+		/// <typeparam name="TResult">The type of the value returned by the function represented by <paramref name="selector"/>.</typeparam>
+		/// <returns>A <see cref="IFutureValue{T}"/>.</returns>
+		/// <exception cref="T:System.ArgumentNullException"><paramref name="source" /> is <see langword="null"/>.</exception>
+		/// <exception cref="T:System.NotSupportedException"><paramref name="source" /> <see cref="IQueryable.Provider"/> is not a <see cref="INhQueryProvider"/>.</exception>
+		public static IFutureValue<TResult> ToFutureValue<TSource, TResult>(this IQueryable<TSource> source, Expression<Func<IQueryable<TSource>, TResult>> selector)
+		{
+			if (source == null)
+			{
+				throw new ArgumentNullException(nameof(source));
+			}
+			if (!(source.Provider is INhQueryProvider provider))
 			{
-				return new FutureValue<T>(() => ((IEnumerable<T>) future));
+				throw new NotSupportedException($"Source {nameof(source.Provider)} must be a {nameof(INhQueryProvider)}");
 			}
 
-			return (IFutureValue<T>) future;
```

Comment:
This line was the implicit `else`, belonging to `ToFutureValue` without selector. The `IFutureValue<T>` case can only happen with an aggregation selector indeed.